op-batcher/batcher/channel_manager.go

```diff
@@ -120,23 +115,23 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
 	s.log.Debug("marked transaction as confirmed", "id", id, "block", inclusionBlock)
 }
 
-// removePendingChannel removes the given completed channel from the manager's state.
-func (s *channelManager) removePendingChannel(channel *channel) {
-	if s.currentChannel == channel {
-		s.currentChannel = nil
-	}
-	index := -1
-	for i, c := range s.channelQueue {
-		if c == channel {
-			index = i
+func (s *channelManager) rewindToBlockWithHash(blockHash common.Hash) {
+	for i, b := range s.blocks {
+		if b.Hash() == blockHash {
+			s.blockCursor = i
 			break
 		}
 	}
-	if index < 0 {
-		s.log.Warn("channel not found in channel queue", "id", channel.ID())
-		return
-	}
-	s.channelQueue = append(s.channelQueue[:index], s.channelQueue[index+1:]...)
+}
+
+// handleChannelTimeout rewinds the channelManager's blockCursor
+// to point at the first block added to the provided channel,
+// and clears the channelQueue and currentChannel.
+func (s *channelManager) handleChannelTimeout(c *channel) {
+	blockHash := c.channelBuilder.blocks[0].Hash()
```

Comment:
Should we check that the the blocks queue is non-empty here? or is the fact that we hit a channel timeout mean that there should have been blocks in the first place?

Reply:
Good call out. 

By design, blocks are only ever enqueued to the channel builder. So the only way for there to be no blocks is if none were ever added. For a channel to timeout, at least one frame must have been submitted, which implies at least one block must have been added to the channel. I think a panic from trying to access a nil slice is next to impossible -- it could happen in an extreme fluke where a different batcher (or previous instance of this batcher) had a colliding channel id, and this batcher happened to have created a channel but not added any blocks to it yet (I'm not sure that can happen before this function runs anyway). 

IMO we need can rely on the uniqueness of channel ids here. If that fails, we probably have bigger problems.